### PR TITLE
feat(entities): 🔥 support id update

### PR DESCRIPTION
**Pull request by DmitryEfimenko**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/elf/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
id update via `updateEntities` is not supported

Issue Number: #146 

## What is the new behavior?
id update via `updateEntities` is supported

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information
The `updateStateIds` function could have been less verbose - I could have used a simple `map`. However, I wanted to optimize the loop and break out from it as soon as ids are updated.

There was no need to update `upsertEntities` since its syntax does not support updating ids. The `upsertEntitiesById` worked out of the box since it uses `updateEntities` internally.

### CODE REVIEW

File: packages/entities/src/lib/update.mutation.ts [hunk 1]
```diff
@@ -47,24 +78,51 @@ export function updateEntities<
...
-  return function (state) {
-    const { ref: { entitiesKey } = defaultEntitiesRef } = options;
+  return function (state, context) {
+    const { ref = defaultEntitiesRef } = options;
     const updatedEntities = {} as Record<
       getIdType<S, Ref>,
       getEntityType<S, Ref>
     >;
 
+    const idProp = getIdKey<keyof getEntityType<S, Ref>>(context, ref);
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:88]: I prefer to go with my first suggestion and pass explicitly this option. There is no need to allocate memory and do all of this if there is no need to.

Add an if statement in line 63 at the original code that takes the updated state and call an updateIds function

Comment (NetanelBasal) [reply packages/entities/src/lib/update.mutation.ts:88]: This isn't a common use case, and I don't want it to affect in any way the original code.

Comment (NetanelBasal) [reply packages/entities/src/lib/update.mutation.ts:88]: I have a better idea

Comment (NetanelBasal) [reply packages/entities/src/lib/update.mutation.ts:88]: We can create a specific mutation:

```ts
store.update(
  updateEntityId(oldId, newId),
  updateEntities(newId, updater)
)
```

Comment (NetanelBasal) [reply packages/entities/src/lib/update.mutation.ts:88]: This will make it tree-shakable and cleaner

Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.ts:88]: If I understand correctly, the purpose of the proposed flag is to get rid of the `changedIds`, which determines if id was changed. Since this use-case is addressing the "optimistic updates" use-case, the `changedIds` will contain a single item 99% of the time. The memory consumption of that object will be minimal.

Besides, even if the flag is introduced, we still need to track what was the old and the new id for each updated entity. I'm not sure how introducing a flag will let us get around that.

Comment (NetanelBasal) [reply packages/entities/src/lib/update.mutation.ts:88]: Never mind. I prefer to go with a new mutation function:

```ts
store.update(
  updateEntitiesIds(oldId, newId), // support single and array
  updateEntities(newId, updater)
)
```

Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.ts:88]: interesting. Looks clean! And should be much simpler to implement too! I'm on it

File: packages/entities/src/lib/update.mutation.ts [hunk 2]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
...
+
+    const { ref = defaultEntitiesRef } = options;
+    const idProp = getIdKey<string>(context, ref);
+    const updatedEntities = { ...state[ref.entitiesKey] };
+
+    for (let i = 0; i < oldIds.length; i++) {
+      const oldVal = oldIds[i];
+      const newVal = newIds[i];
+
+      if (state[ref.entitiesKey][newVal]) {
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:286]: Let's remove this protection, it's redundant.

File: packages/entities/src/lib/update.mutation.ts [hunk 3]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
...
+      }
+
+      const oldEntity = state[ref.entitiesKey][oldVal];
+      const updated = { ...oldEntity, [idProp]: newVal };
+
+      updatedEntities[newVal] = updated;
+      Reflect.deleteProperty(updatedEntities, oldVal);
+    }
+
+    const updatedStateIds: getIdType<S, Ref>[] = [...state[ref.idsKey]];
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:299]:
```suggestion
    const updatedStateIds: getIdType<S, Ref>[] = state[ref.idsKey].slice()
```

File: packages/entities/src/lib/update.mutation.ts [hunk 4]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
     };
   };
 }
+
+/**
+ * Update entities ids
+ *
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:249]: Update the docs, please

File: packages/entities/src/lib/update.mutation.ts [hunk 5]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
...
+ * @example
+ *
+ * // update single entity id
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:252]:
```suggestion
 * // Update a single entity id
```

File: packages/entities/src/lib/update.mutation.ts [hunk 6]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
...
+ * store.update(updateEntitiesIds(1, 2));
+ *
+ * // update multiple entities ids
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:255]:
```suggestion
 * // Update multiple entities ids
```

File: packages/entities/src/lib/update.mutation.ts [hunk 7]
```diff
@@ -243,3 +243,85 @@ export function upsertEntities<
...
+ * store.update(updateEntitiesIds([1, 2], [10, 20]));
+ *
+ * // update entity id using a custom ref
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:258]:
```suggestion
 * // Update entity id using a custom ref
```

File: packages/entities/src/lib/update.mutation.ts (hunk 3 above)
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:299]:
```suggestion
    const currentIds: getIdType<S, Ref>[] = [...state[ref.idsKey]];
```

File: packages/entities/src/lib/update.mutation.spec.ts [hunk 8]
```diff
@@ -203,4 +190,35 @@ describe('update', () => {
...
+      expect(() => {
+        store.update(updateEntitiesIds(1, 2));
+      }).toThrow();
+    });
+
+    it('should work with ref', () => {
+      const store = createUIEntityStore();
+      store.update(addEntities([createUITodo(1)], { ref: UIEntitiesRef }));
+      toMatchSnapshot(expect, store, 'id updated false');
+      store.update(updateEntitiesIds(1, 2, { ref: UIEntitiesRef }));
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.spec.ts:220]: add a test that updated the entity after changing the id, please

File: packages/entities/src/lib/update.mutation.spec.ts [hunk 9]
```diff
@@ -164,23 +164,10 @@ describe('update', () => {
       const todo = createTodo(1);
       store.update(addEntities([todo]));
 
-      // update the todo
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.spec.ts:167]: why removed?

File: packages/entities/src/lib/__snapshots__/update.mutation.spec.ts.snap [hunk 10]
```diff
@@ -91,7 +188,7 @@ Object {
     "2": Object {
       "completed": false,
       "id": 2,
-      "title": "elf",
+      "title": "todo 2",
```
Comment (NetanelBasal) [inline packages/entities/src/lib/__snapshots__/update.mutation.spec.ts.snap:191]: why the change?

File: packages/entities/src/lib/update.mutation.ts (hunk 2 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.ts:286]: This scenario is hopefully rare, but if the code does not throw, it could lead to some nasty results.
Imagine a scenario:
```ts
// the user wants to update id 2 to become 3 and then id 1 to become 2
// the following would work
store.update(
  updateEntitiesIds(2, 3),
  updateEntitiesIds(1, 2)
);
```
however, if they have code like below...
```ts
store.update(updateEntitiesIds([1, 2], [2, 3]));
```
... the operation would succeed and this would be the store value:

```ts
{
  entities: { '3': { id: 3, title: 'todo 2', completed: false } },
  ids: [ 2, 3 ]
}
```
As you can see, the store is out of control.
I, as a developer, would be happy to see this error and figure out why my code is trying to update to ids that already exist.

File: packages/entities/src/lib/__snapshots__/update.mutation.spec.ts.snap (hunk 10 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/__snapshots__/update.mutation.spec.ts.snap:191]: The diff can't quite figure out what the change was. It's hard to review here.
here's an actual change (removed work "update" - I think it was a typo):
```
exports[`update UpsertById should update add the missing entities and update existing: two entities, 1: title "elf", 2: title "todo 2" 1`] = `
exports[`update UpsertById should add the missing entities and update existing: two entities, 1: title "elf", 2: title "todo 2" 1`] = `
```

File: packages/entities/src/lib/update.mutation.spec.ts (hunk 9 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.spec.ts:167]: The next two lines [were](https://github.com/ngneat/elf/blob/master/packages/entities/src/lib/update.mutation.spec.ts#L168-L169):
```ts
todo.completed = true;
store.update(upsertEntities([todo]));
```
Even if you would comment out the `store.update()` line, the test would still pass since we're mutating the todo directly in the store. After changing it to the immutable version I noticed that this test became exactly the same as the next one.

File: packages/entities/src/lib/update.mutation.spec.ts (hunk 8 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.spec.ts:220]: will do

File: packages/entities/src/lib/update.mutation.ts (hunk 4 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.ts:249]: sure thing. I wanted to make sure the code looks good before starting with the docs.

File: packages/entities/src/lib/update.mutation.ts (hunk 3 above)
Comment (DmitryEfimenko) [reply packages/entities/src/lib/update.mutation.ts:299]: except that these are not current ids. We updated the reference to the array and we'll be changing some of the ids in that array too. Later on, when we return updated state, the `currentIds` looks weird:
```ts
return {
  ...state,
  [ref.entitiesKey]: updatedEntities,
  [ref.idsKey]: currentIds,
};
```

**Comment by DmitryEfimenko** [comment]
Thanks for merging it. Looking forward to seeing it released on npm!
